test/integration/scheduler/volume_binding_test.go

```diff
@@ -0,0 +1,405 @@
+						{
+							Key:      labelKey,
+							Operator: v1.NodeSelectorOpIn,
+							Values:   []string{labelValue},
+						},
+					},
+				},
+			},
+		},
+	}
+)
+
+func TestVolumeBinding(t *testing.T) {
+	config := setup(t)
+	defer config.teardown()
+
+	scs := []*storagev1.StorageClass{
+		makeStorageClass(classWait, &modeWait),
+		makeStorageClass(classImmediate, &modeImmediate),
+	}
+
+	// Each test case should use unique API objects because
+	// etcd is not restarted in between each test
+	cases := map[string]struct {
+		pod  *v1.Pod
+		pvs  []*v1.PersistentVolume
+		pvcs []*v1.PersistentVolumeClaim
+	}{
+		"immediate can bind": {
+			pod:  makePod("pod-i-canbind", config.ns, []string{"pvc-i-canbind"}),
+			pvs:  []*v1.PersistentVolume{makePV(t, "pv-i-canbind", classImmediate, "", "")},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-i-canbind", config.ns, &classImmediate, "")},
+		},
+		"immediate pvc prebound": {
+			pod:  makePod("pod-i-pvc-prebound", config.ns, []string{"pvc-i-prebound"}),
+			pvs:  []*v1.PersistentVolume{makePV(t, "pv-i-pvc-prebound", classImmediate, "", "")},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-i-prebound", config.ns, &classImmediate, "pv-i-pvc-prebound")},
+		},
+		"immediate pv prebound": {
+			pod:  makePod("pod-i-pv-prebound", config.ns, []string{"pvc-i-pv-prebound"}),
+			pvs:  []*v1.PersistentVolume{makePV(t, "pv-i-prebound", classImmediate, "pvc-i-pv-prebound", config.ns)},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-i-pv-prebound", config.ns, &classImmediate, "")},
+		},
+		"wait can bind": {
+			pod:  makePod("pod-w-canbind", config.ns, []string{"pvc-w-canbind"}),
+			pvs:  []*v1.PersistentVolume{makePV(t, "pv-w-canbind", classWait, "", "")},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-w-canbind", config.ns, &classWait, "")},
+		},
+		"wait pvc prebound": {
+			pod:  makePod("pod-w-pvc-prebound", config.ns, []string{"pvc-w-prebound"}),
+			pvs:  []*v1.PersistentVolume{makePV(t, "pv-w-pvc-prebound", classWait, "", "")},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-w-prebound", config.ns, &classWait, "pv-w-pvc-prebound")},
+		},
+		"wait pv prebound": {
+			pod:  makePod("pod-w-pv-prebound", config.ns, []string{"pvc-w-pv-prebound"}),
+			pvs:  []*v1.PersistentVolume{makePV(t, "pv-w-prebound", classWait, "pvc-w-pv-prebound", config.ns)},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-w-pv-prebound", config.ns, &classWait, "")},
+		},
+		"wait can bind two": {
+			pod: makePod("pod-w-canbind-2", config.ns, []string{"pvc-w-canbind-2", "pvc-w-canbind-3"}),
+			pvs: []*v1.PersistentVolume{
+				makePV(t, "pv-w-canbind-2", classWait, "", ""),
+				makePV(t, "pv-w-canbind-3", classWait, "", ""),
+			},
+			pvcs: []*v1.PersistentVolumeClaim{
+				makePVC("pvc-w-canbind-2", config.ns, &classWait, ""),
+				makePVC("pvc-w-canbind-3", config.ns, &classWait, ""),
+			},
+		},
+		// TODO: this testcase is failing with pod cache modification error
```

Comment:
I need to debug this failing testcase

Reply:
Found it, needed to make a copy of the assumedPod object before updating it.

Reply:
Yes! I am glad that the test caught the issue.